test/integration/short/prepare-simulator-tests.js

```diff
@@ -0,0 +1,166 @@
+              var queryLog = logs[i];
+              if (queryLog.type === "PREPARE" && queryLog.query === query) {
+                prepareQuery = queryLog;
+              }
+            }
+            if (!prepareQuery) {
+              assert.fail('Query no prepared on all hosts');
+            }
+            next();
+          });
+        }, done);
+      });
+    });
+    it('should re-prepare query when host go UP again', function (done) {
+      var idRandom = types.Uuid.random();
+      var nodeDownAddress = sCluster.getContactPoints()[4];
+      utils.series(
+        [
+          function stopLastNode(next) {
+            sCluster.stopNode(nodeDownAddress, next);
+          },
+          function runQuery(next) {
+            utils.timesSeries(5, function (n, nextIteration) {
+              client.execute(query, [idRandom], {prepare: 0}, function (err, result) {
+                assert.ifError(err);
+                assert.strictEqual(client.hosts.length, 5);
+                assert.notEqual(result, null);
+                assert.notEqual(result.rows, null);
+                nextIteration();
+              });
+            }, next);
+          },
+          function verifyIfNodeIsMarkedDown(next) {
+            var nodeDown = client.hosts.get(nodeDownAddress);
+            nodeDown.checkIsUp();
+            assert(!nodeDown.isUp());
+            next();
+          },
+          function prepareQuery(next) {
+            client.execute(query, [idRandom], {prepare: 1}, function (err, result) {
+              assert.ifError(err);
+              assert.strictEqual(client.hosts.length, 5);
+              assert.notEqual(result, null);
+              assert.notEqual(result.rows, null);
+              next();
+            });
+          },
+          function verifyLogs(next) {
+            utils.eachSeries(client.hosts.values(), function(host, nextHost) {
+              sCluster.queryNodeLog(host.address, function(logs) {
+                var prepareQuery;
+                for(var i = 0; i < logs.length; i++) {
+                  var queryLog = logs[i];
+                  if (queryLog.type === "PREPARE" && queryLog.query === query) {
+                    prepareQuery = queryLog;
+                  }
+                }
+                if (!prepareQuery) {
+                  assert.strictEqual(nodeDownAddress, host.address);
+                } else {
+                  assert.notEqual(prepareQuery, undefined);
+                }
+                nextHost();
+              });
+            }, next);
+          },
+          function resumeLastNode(next) {
+            var nodeDown = client.hosts.get(nodeDownAddress);
+            nodeDown.on('up', function() {
+              helper.trace("Node marked as UP");
+              setTimeout(next, 1000); //give time for driver to re prepare statement
+            });
+            sCluster.resumeNode(nodeDownAddress, function() {
+              nodeDown.checkIsUp();
```

Comment:
We should avoid using internal methods in the integration tests (whenever possible).